Plutarch/List.hs

```diff
@@ -0,0 +1,277 @@
+-- | Scott-encoded lists and ListLike typeclass
+module Plutarch.List (
+  PScottList (..),
+  PListLike (..),
+  pconvertLists,
+
+  -- * Comparison
+  plistEquals,
+
+  -- * Query
+  pelem,
+  plength,
+  pnull,
+
+  -- * Construction
+  psingleton,
+
+  -- * Combine
+  pconcat,
+  pzipWith,
+  pzipWith',
+  pzip,
+
+  -- * Traversals
+  pmap,
+  pfilter,
+
+  -- * Catamorphisms
+  precList,
+  pfoldr,
+  pfoldr',
+  pall,
+) where
+
+import Plutarch
+import Plutarch.Bool (PBool (..), PEq (..), pif, (#&&))
+import Plutarch.Integer (PInteger)
+import Plutarch.Pair (PPair (..))
+import Plutarch.Prelude
+
+import Data.Kind
+
+data PScottList (a :: k -> Type) (s :: k)
```

Comment:
We have to think about naming. `PMaybe` is scott-encoded, and #66 it adding Data-encoded `PMaybe` for use within datums and similar.

Reply:
Currently no prefix has been Scott-encoded, and `Builtin` has been used as a prefix for the Data-encoded ones, but that doesn't fit for e.g. `PScriptContext`. Perhaps we want to have some explicit relation between the Scott-encoded and Data-encoded variants.

Reply:
Yeah, I suppose `PList` is fine for _this_ case, I'll rename.